app/Http/Controllers/DocuSignController.php

```diff
@@ -97,71 +144,312 @@ public function signAgreement(Request $request)
+        $tokens = $docusign->generateAccessToken(
+            client_id: config('docusign.client_id'),
+            client_secret: config('docusign.client_secret'),
+            code: $request->code
+        )[0];
 
-                config(
-                    'docusign.membership_agreement.archive_mailbox_name'
-                ).'_UserName' => 'Membership Agreement Archives',
+        /** @var \DocuSign\eSign\Client\Auth\UserInfo $userinfo */
+        $userinfo = $docusign->getUserInfo($tokens->getAccessToken())[0];
 
-                config(
-                    'docusign.membership_agreement.archive_mailbox_name'
-                ).'_Email' => config('services.membership_agreement_archive_email'),
-            ]
-        );
+        $userInSameAccount = false;
+
+        /** @var \DocuSign\eSign\Client\Auth\Account $account */
+        foreach ($userinfo->getAccounts() as $account) {
+            if ($account->getAccountId() === config('docusign.account_id')) {
+                $userInSameAccount = true;
+            }
+        }
+
+        if (! $userInSameAccount) {
+            $userinfo_serialized = DocuSign::serializeUserInfo($userinfo);
+
+            Log::info($userinfo_serialized);
+
+            abort(401);
+        }
+
+        switch ($request->session()->get('next')) {
+            case 'getGlobalToken':
+                // In this case, we just need to output the user ID so that it can be stored in the environment.
+                // This doesn't do anything interesting in DocuSign, it's just confirmation that the app has access to
+                // impersonate a user at this point.
+
+                $userinfo_serialized = DocuSign::serializeUserInfo($userinfo);
+
+                Log::info($userinfo_serialized);
+
+                DocuSign::getApiClient();
+
+                return response()->json($userinfo_serialized);
+            case 'signAgreement':
+                // In this case, a user has just authenticated with DocuSign after starting the signing flow.
+                // We need to store the credentials in their user model, then continue with signing.
+
+                /** @var \App\Models\User $user */
+                $user = $request->user();
+
+                if ($userinfo->getEmail() !== $user->uid.'@gatech.edu') {
+                    $userinfo_serialized = DocuSign::serializeUserInfo($userinfo);
+
+                    Log::info($userinfo_serialized);
+
+                    abort(401);
+                }
+
+                $user->docusign_access_token = $tokens->getAccessToken();
+                $user->docusign_access_token_expires_at = Carbon::now()->addSeconds($tokens->getExpiresIn() - 60);
+                $user->docusign_refresh_token = $tokens->getRefreshToken();
+                $user->docusign_refresh_token_expires_at = Carbon::now()->addDays(29);
+                $user->save();
+
+                return $this->signAgreement($request);
+            default:
+                abort(500);
```

Comment:
Feels like we should add some kind of error/log here rather than just throwing an empty 500/